Add `EVM` methods for `encodeABI` & `decodeABI`

m-Peter opened the pull request:
Closes https://github.com/onflow/flow-go/issues/4937

Introduces two natively-implemented methods on the `EVM` contract, for encoding/decoding ABI:
- `EVM.encodeABI`
- `EVM.decodeABI`

**Note:** Some stuff are still to be decided.

turbolent reviewed (commented):
Great work!

As discussed on Discord, it might be nice to "future-proof" the Cadence API for the EVM ABI, so that it can be easily extended without breaking changes in the future.

Proposal: Make the API more object-oriented:
- Introduce a Cadence type for the ABI, and allow creating an instance of it from a JSON string. Later this can be extended to allow constructing an ABI without the JSON.
- Refactor the current `encodeABI/decodeABI` functions of the `EVM` contract into functions of the ABI type. This can be later used to encode arguments even for an ABI that was not constructed from JSON.

This basically reflects geth's API, i.e. [`abi.ABI`](https://github.com/ethereum/go-ethereum/blob/5bf8769fb081c5f17f23b1b55e508789f0a8e632/accounts/abi/abi.go#L34C18-L34C18). As mentioned above, later we could introduce ABI types like `EVM.Method`, `EVM.Event`, etc. and have getter functions on the ABI type (again, similar to the Go API).

Concretely, maybe we can have something like 

```cadence
contract EVM {

    // ...

    struct ABI {
        fun encode(arguments: [AnyStruct]): [UInt8] { /* ... */ }
        fun decode(data: [UInt8]): [AnyStruct] { /* ... */ }
    }

    fun abiFromJSON(): ABI { /* ... */ }
```

This is just an idea, and maybe out of scope for the MVP, happy to get this in and we can discuss the re-design and implement it if necessary in a follow-up PR.

ramtinms reviewed (approved):
Looks good to me, we just need to add more comprehensive computation metering here, 

Maybe not 100% related to this PR but, I'm a bit confused about the application, if the encoding data is known in advance and this is a pure function, why do we need to do it on chain?

turbolent commented:
@franklywatson Do you have any more information as to how cross-chain integration work needs the Ethereum ABI encoding and decoding functions to be available on-chain?

ryan-suematsu-anchain commented:
> @franklywatson Do you have any more information as to how cross-chain integration work needs the Ethereum ABI encoding and decoding functions to be available on-chain?

Hey @turbolent 👋 , 

I am one of the devs for the axelar-flow bridge.  Heres a quick run down of ABI encode/decode and why we need it for our use case.

So EVM frequently uses encode/decode abi to pass data between on chain smart contracts and dapps, sometimes thats a way to return dynamic data, or pass in dynamic function args to functions.   It can be thought of as a way to 'stringify'/serialize your data in a portable format.

In our case we need it to pass dynamic data into dapp functions after the data has been signed by the axelar network.  The flow goes something like this,

- the axelar network validates a GMP (function call)

- the axelar network then signs the gmp data by creating a hash of the input data.  This input data is something like, (destinationAddr, sourceAddr, destinationChain, sourceChain, payload).  (The payload is what is key here).  This payload is generated by the source dapp, often created on the corresponding chain.

          -  This payload needs to be a string of data that can be signed off chain, passed on chain, then parsed to its required arguments.

- This hash gets passed to the cross chain service, where we can validate the hash was signed by the correct individuals and we can recreate the hash to verify that the data passed in was indeed the data that was signed.

- ...some internal logic happens to approve the call

- When that call to the destination contract (on flow) is made, we pass in the payload as well and do the same logic checks. to ensure that the payload and all concerning data was included in the signed hash.

- after that is verified, the payload is then passed to the destination function, where the data can be decoded/parsed into its required format.

its crucial that we use some sort of arbitrary serialization technique for the payload, as we want to allow the destination dapp to be able to pass in as many custom arguments as required.
its also good to note, that EVM doesnt use something like json parsing and stringifying because string manipulation is very expensive on chain.

sideninja commented on fvm/evm/stdlib/contract_test.go:
```diff
@@ -236,6 +236,298 @@ func newEVMScriptEnvironment(handler types.ContractHandler, service flow.Address
+
+	encodedABI := cadence.NewArray(
+		[]cadence.Value{
+			cadence.UInt8(0xef), cadence.UInt8(0x84), cadence.UInt8(0xad), cadence.UInt8(0x63),
```
maybe write a function that expands the values, so something like:
pseudo:
```
cadenceUInt8(vals ...cadence.UIn8) []cadence.UInt8 {
 for 32 - len(vals) ->  append vals []UInt8{0x0}
 return vals
}
```
so you generate extended array

turbolent commented:
@ryan-suematsu-anchain Thank you for the great write-up Ryan!

Just to confirm, would the functions as proposed/implemented in the PR be useful for you? If I understand you correctly, at the end of your comment you're saying that passing the ABI to the chain to encode values to data, or decode encoded data to values, is uncommon (not done at all?). 

In the last point, you mention "the data can be decoded/parsed into its required format". Would that be the place where the ABI decoding function would be called?

franklywatson commented:
@turbolent was reading your comment "you're saying that passing the ABI to the chain to encode values to data, or decode encoded data to values, is uncommon (not done at all?)". The idea that because it's a rarely used set of functionality it's maybe not that useful isn't the right question here IMO. 

To my understanding, the need for the functionality arises from the fact that Flow's Axelar gateway contracts are merely one node on the cross-chain network messaging protocol. The expectations of the protocol is that participating gateways must support this since it does get used (albeit infrequently) and since it's decentralized we need to have parity with the rest of the network.

ryan-suematsu-anchain commented:
@franklywatson is right, I think its an important feature to have not only for us but other cross chain applications looking to build for Flow/EVM cross chain solutions.

@turbolent I think there might be a misunderstanding regarding the difference between something like Json.parse/stringify and ABI.encode/decode.  In EVM the abi encoding takes advantage of alot of optimizations since its a byte encoding format, where as a json stringify/parse format uses string manipulation and often avoided because of the high gas cost that comes with it.  I would say ABI encode/decode is pretty frequently in various EVM cases such as, complicated return structs, passing in dynamic lengths of multiple args, and cross chain apps!

"Just to confirm, would the functions as proposed/implemented in the PR be useful for you?" 
Yes!!

"In the last point, you mention "the data can be decoded/parsed into its required format". Would that be the place where the ABI decoding function would be called?"
Yes!!

turbolent commented:
@franklywatson @ryan-suematsu-anchain Thank you for the additional info and answers 👍

Follow-up question: Given that encoding and decoding is useful on-chain, looking at e.g. https://www.wtf.academy/en/solidity-advanced/ABIEncode/, how common is it to pass the ABI for a _whole contract_ on-chain , and use it to encode/decode, vs. what the linked page describes, encoding arguments / decoding encoded arguments given a _single function signature_?

For example, we might also want to have, or even only have, encode/decode functions like, e.g. 

```cadence
contract EVM {
  fun encode(signature: String, arguments: [AnyStruct]): [UInt8] { /* ... */ }
  fun decode(signature: String, data: [UInt8]): [AnyStruct] { /* ... */ }
}

let data = EVM.encode(signature: "foo(uint256,string)", arguments: [42, "bar"])
let arguments = EVM.decode(signature: "foo(uint256,string)", data)
```

Do we need/want packed encoding?

turbolent commented on fvm/evm/stdlib/contract.go:
```diff
@@ -38,6 +40,180 @@ var evmAddressBytesType = sema.NewConstantSizedType(nil, sema.UInt8Type, types.A
+					arguments = append(arguments, uint64(value))
+				}
+			}
+			packed, err := abi.Pack("details", arguments...)
```
Now that I understand the ABI encoding/decoding more: Doesn't this hardcode the method name to "details"? Shouldn't the method name be passed to the `encodeABI` (and `decodeABI`) function, i.e. should those functions have an additional parameter?

ryan-suematsu-anchain replied:
https://github.com/axelarnetwork/axelar-gmp-sdk-solidity/blob/66325132bde3f61d2c18748cfcd3e181b063ea00/contracts/test/mocks/MockGateway.sol#L224C32-L224C32

Here is a good example of how we plan to use ABI encode/decode on chain.  Mainly passing in a 'partial' abi to encode only those variables into bytes.  I would say encode packed is less important of a feature as it seems to do everything normal encode/decode does in less space.  It may be good to support in a future roadmap.  I think the functions you outlined are a good place to start.  It would allow developers to create data packages that would be portable between flow and evm as well as generate EVM function calls/signatures on flow.

turbolent commented:
We had a call about several details of the API

## Functionality

Solidity provides several different functions for encoding and decoding data (e.g. see https://www.wtf.academy/en/solidity-advanced/ABIEncode/). For example, it supports
  1. Encoding of just multiple values, either "packed" (shorter) or unpacked (`abi.encode` and `abi.encodePacked`)
  2. Encoding of function call, by passing function name and arguments (`abi.encodeWithSignature`)

To start with, we need at least 1), and the packed variant is desirable.
In the future we might also want to consider adding Cadence equivalents of the other ABI encoding/decoding functions.

If the Cadence encoding function produces a packed encoding, it should probably indicate that in the name, e.g `encodePacked`, so that later we can also add e.g. a encoding function for the unpacked variant.

## Cadence API

For example, the Cadence API could make use of type parameters, and similar to the Solidity function, take all values through a variable-sized parameter list (straw man syntax, as Cadence syntax does not allow such functions to be written):

```cadence
fun encodePacked<T1, ..., TN>(_ v1: T1, ..., _ vN: TN): [UInt8]
```

Essentially, this function has a variable length of parameters / accepts an arbitrary number of arguments, and has one type parameter for each (value) parameter.

This would allow a developer to simply call the function with any arguments, and the type checker would infer the type parameters automatically:

For example, this function call would encode an `Int` and a `String` (inferred from the arguments):

```cadence
let data = encodePacked(42, "bar")
```

If however the user wanted to encode the integer literal as a `UInt256`, they could do so by either asserting the type in the argument expression, or enforce the type in the type parameter list:

```cadence
let data = encodePacked(42 as UInt256, "bar")
```

```cadence
let data = encodePacked<UInt256, String>(42, "bar")
```

Similarly, the decoding function could look like:

```cadence
fun decode<T1, ..., TN>(data: [UInt8]): [U]
```

Note that given that Cadence does not (yet) support multiple return values or tuples, the return type would be `[U]`, where `U` is the least-common supertype of all decoded types (`T1`, ..., `TN`).

A function call could look like:

```cadence
let values = EVM.decode<UInt256, String>(data)   // `values` has type `[AnyStruct]`
```

## Mapping between Cadence and Solidity Types

The mapping between Solidity types and Cadence types could look like
- `bool` <-> `Bool`
- `string` <-> `String`
- Integer types: `uint256` <-> `UInt256`. For sizes where there is no Cadence equivalent: maybe just `Int` or `UInt`?
- Array
- Address: 
  - `address` decodes to `EVM.EVMAddres`
  - Allow encoding of `String` and `EVM.EVMAddres` to `address`

## Metering

All computation performed by encoding and decoding needs to be metered. That means that the native (Go) code must be annotated with computation metering calls. The metering needs to handle containers (arrays). 

Example: RLP encoding/decoding

m-Peter commented on fvm/evm/stdlib/contract.go:
```diff
@@ -38,6 +40,272 @@ var evmAddressBytesType = sema.NewConstantSizedType(nil, sema.UInt8Type, types.A
+					}
+					arguments = append(arguments, gethABI.Argument{Type: typ})
+				case *interpreter.CompositeValue:
+					if value.QualifiedIdentifier == "EVM.EVMAddress" {
```
This check should probably be more bulletproof. Any idea on how to match the desired type `EVM.EVMAddress`?

m-Peter commented on fvm/evm/stdlib/contract.go:
```diff
@@ -38,6 +40,272 @@ var evmAddressBytesType = sema.NewConstantSizedType(nil, sema.UInt8Type, types.A
+			values := make([]interface{}, 0)
+			var arguments gethABI.Arguments
+
+			for i := 0; i < valuesArray.Count(); i++ {
```
I could also use `valuesArray.Iterate` if it is preferable.

m-Peter replied:
Updated in https://github.com/onflow/flow-go/pull/5024/commits/e82de52f495f2b00b592494c58288d403dd61453

m-Peter commented on fvm/evm/stdlib/contract.go:
```diff
@@ -38,6 +40,272 @@ var evmAddressBytesType = sema.NewConstantSizedType(nil, sema.UInt8Type, types.A
+					if value.QualifiedIdentifier == "EVM.EVMAddress" {
+						bytes, err := interpreter.ByteArrayValueToByteSlice(
+							inter,
+							value.GetMember(inter, locationRange, "bytes"),
```
This "bytes" should not remain hardcoded, should use the appropriate constant, if there's any already.

m-Peter replied:
Updated in https://github.com/onflow/flow-go/pull/5024/commits/a483dc78245b76912702efade3a9f0233b2939a4

m-Peter commented on fvm/evm/stdlib/contract.go:
```diff
@@ -38,6 +40,272 @@ var evmAddressBytesType = sema.NewConstantSizedType(nil, sema.UInt8Type, types.A
+						arguments = append(arguments, gethABI.Argument{Type: typ})
+					}
+				case *interpreter.ArrayValue:
+					elements := make([]uint64, 0)
```
We should also decide on which primitives we are going to support for encoding/decoding arrays. Here I have just added support for `[UInt64]`

m-Peter commented on fvm/evm/stdlib/contract.go:
```diff
@@ -38,6 +40,272 @@ var evmAddressBytesType = sema.NewConstantSizedType(nil, sema.UInt8Type, types.A
+			Label:      sema.ArgumentLabelNotRequired,
+			Identifier: "types",
+			TypeAnnotation: sema.NewTypeAnnotation(
+				sema.NewVariableSizedType(nil, sema.AnyType),
```
I was not sure what `TypeAnnotation` to use for the `types` argument, given the current definition:

```cadence
access(all)
fun decodeABI(_ types: [Type], data: [UInt8]): [AnyStruct] {
    return InternalEVM.decodeABI(types, data: data)
}
```
I am probably not using the correct one :thinking:

m-Peter replied:
Updated in https://github.com/onflow/flow-go/pull/5024/commits/22ab4ae59ab0bf715cbf4751d136c6c92c782d3d

m-Peter commented on fvm/evm/stdlib/contract.go:
```diff
@@ -38,6 +40,272 @@ var evmAddressBytesType = sema.NewConstantSizedType(nil, sema.UInt8Type, types.A
+			for i := 0; i < typesArray.Count(); i++ {
+				arg := typesArray.Get(inter, locationRange, i)
+				switch arg.String() {
+				case "Type<String>()":
```
I didn't find any better way than this ugly:

```go
switch arg.String() {
case "Type<String>()":
```

But I am pretty sure there's a better way to do this.

m-Peter replied:
Updated in https://github.com/onflow/flow-go/pull/5024/commits/22ab4ae59ab0bf715cbf4751d136c6c92c782d3d

m-Peter commented on fvm/evm/stdlib/contract.go:
```diff
@@ -38,6 +40,272 @@ var evmAddressBytesType = sema.NewConstantSizedType(nil, sema.UInt8Type, types.A
+				case gethCommon.Address:
+					var address types.Address
+					copy(address[:], value.Bytes())
+					values = append(values, EVMAddressToAddressBytesArrayValue(inter, address))
```
Here I should actually return the whole `EVM.EVMAddress` struct, and not just its `bytes`. I didn't find a way how to create it though.

m-Peter commented on fvm/evm/stdlib/contract_test.go:
```diff
@@ -236,6 +236,237 @@ func newEVMScriptEnvironment(handler types.ContractHandler, service flow.Address
+        var data = EVM.encodeABI([address])
+        var values = EVM.decodeABI([Type<EVM.EVMAddress>()], data: data)
+        assert(values.length == 1)
+        assert((values[0] as! [UInt8; 20]) == address.bytes)
```
The `EVM.decodeABI` for `Type<EVM.EVMAddress>()` should actually return a `EVM.EVMAddress` struct, and not just the `bytes`.

m-Peter commented on fvm/evm/stdlib/contract.go:
```diff
@@ -38,6 +40,272 @@ var evmAddressBytesType = sema.NewConstantSizedType(nil, sema.UInt8Type, types.A
+				switch value := element.(type) {
+				case *interpreter.StringValue:
+					values = append(values, value.Str)
+					typ, err := gethABI.NewType("string", "", nil)
```
Good point :+1: They can be reused across multiple calls, which means we can move most of them in top-level `var`s. Except for maybe slices, as we cannot possibly list all of their sizes, e.g.:
```go
typ, err := getABI.NewType("uint64[20]", "", nil)
typ, err = getABI.NewType("uint64[15]", "", nil)
```

m-Peter commented on fvm/evm/stdlib/contract.go:
```diff
@@ -38,6 +40,272 @@ var evmAddressBytesType = sema.NewConstantSizedType(nil, sema.UInt8Type, types.A
+var internalEVMTypeDecodeABIFunctionType = &sema.FunctionType{
+	Parameters: []sema.Parameter{
+		{
+			Label:      sema.ArgumentLabelNotRequired,
```
Updated in https://github.com/onflow/flow-go/pull/5024/commits/22ab4ae59ab0bf715cbf4751d136c6c92c782d3d

m-Peter commented on fvm/evm/stdlib/contract.go:
```diff
@@ -38,6 +40,272 @@ var evmAddressBytesType = sema.NewConstantSizedType(nil, sema.UInt8Type, types.A
+						panic(err)
+					}
+					arguments = append(arguments, gethABI.Argument{Type: typ})
+				case interpreter.UInt64Value:
+					values = append(values, uint64(value))
+					typ, err := gethABI.NewType("uint64", "", nil)
```
@turbolent I have added almost of the `uint`/`int` numeric types in https://github.com/onflow/flow-go/pull/5024/commits/94947a13a9382a0d3ba3d2686bf8c5652cebe5ae
I am unsure about the encoding/decoding of `UInt128` / `UInt256` / `Int128` / `Int256`. How do I handle these in https://github.com/onflow/flow-go/pull/5024/commits/94947a13a9382a0d3ba3d2686bf8c5652cebe5ae#diff-a7da8c11cf2ee1e30d5a9d210284cbdf9b44f30838d922ee25033b83336fd703R445-R451 ?

turbolent replied:
Will reply there

turbolent commented on fvm/evm/stdlib/contract.go:
```diff
@@ -38,6 +42,491 @@ var evmAddressBytesType = sema.NewConstantSizedType(nil, sema.UInt8Type, types.A
+						}
+						elements = append(elements, uint64(v))
+
+						return true
```
```suggestion
						// continue iteration
						return true
```

turbolent commented on fvm/evm/stdlib/contract.go:
```diff
@@ -38,6 +42,491 @@ var evmAddressBytesType = sema.NewConstantSizedType(nil, sema.UInt8Type, types.A
+						panic(err)
+					}
+					arguments = append(arguments, gethABI.Argument{Type: typ})
+				}
```
Add a default case and panic with a new user error that indicates the passed value is not supported / cannot be encoded. Currently the switch is silently ignoring unsupported values. Maybe also add test cases

turbolent commented on fvm/evm/stdlib/contract.go:
```diff
@@ -38,6 +42,491 @@ var evmAddressBytesType = sema.NewConstantSizedType(nil, sema.UInt8Type, types.A
+			values := make([]interface{}, 0)
+			var arguments gethABI.Arguments
+
+			valuesArray.Iterate(inter, func(element interpreter.Value) (resume bool) {
```
This inner function, which converts an `interpreter.Value` to a `gethABI.Argument`.
Maybe extract it into a global, self-standing function, e.g.

```go
func convertValueToGethArgument(value interpreter.Value) (getABI.Argument, error) {
	// ...
}
```

turbolent commented on fvm/evm/stdlib/contract.go:
```diff
@@ -38,6 +42,491 @@ var evmAddressBytesType = sema.NewConstantSizedType(nil, sema.UInt8Type, types.A
+							panic(err)
+						}
+						arguments = append(arguments, gethABI.Argument{Type: typ})
+					}
```
Report an error if the value is not an `EVM.EVMAddress` instead of silently ignoring the composite value

turbolent commented on fvm/evm/stdlib/contract.go:
```diff
@@ -38,6 +42,491 @@ var evmAddressBytesType = sema.NewConstantSizedType(nil, sema.UInt8Type, types.A
+						v, ok := element.(interpreter.UInt64Value)
+						if !ok {
+							panic(
+								fmt.Errorf(
```
This should become a new error type (user error), e.g. something like "`unsupportedValueABIEncodingError`"

turbolent commented on fvm/evm/stdlib/contract.go:
```diff
@@ -38,6 +42,491 @@ var evmAddressBytesType = sema.NewConstantSizedType(nil, sema.UInt8Type, types.A
+					panic(errors.NewUnreachableError())
+				}
+
+				switch value := typeValue.Type.(type) {
```
```suggestion
				switch typeValue := typeValue.Type.(type) {
```

turbolent commented on fvm/evm/stdlib/contract.go:
```diff
@@ -38,6 +42,491 @@ var evmAddressBytesType = sema.NewConstantSizedType(nil, sema.UInt8Type, types.A
+				}
+
+				switch value := typeValue.Type.(type) {
+				case interpreter.ArrayStaticType:
+					typ, err := gethABI.NewType("uint64[]", "", nil)
```
Check the array static type's element type

turbolent commented on fvm/evm/stdlib/contract.go:
```diff
@@ -38,6 +42,491 @@ var evmAddressBytesType = sema.NewConstantSizedType(nil, sema.UInt8Type, types.A
+				case interpreter.ArrayStaticType:
+					typ, err := gethABI.NewType("uint64[]", "", nil)
+					if err != nil {
+						panic(err)
```
Like for encoding: The geth errors need to be mapped to user errors.

turbolent commented on fvm/evm/stdlib/contract.go:
```diff
@@ -38,6 +42,491 @@ var evmAddressBytesType = sema.NewConstantSizedType(nil, sema.UInt8Type, types.A
+
+			values := make([]interpreter.Value, 0)
+			for _, arg := range decoded {
+				switch value := arg.(type) {
+				case string:
```
Instead of switching over the Go type of the decoded value, look up the Cadence type in `typesArray`, and decode based on the Cadence type. Maybe even switch the loop from iterating over the decoded values to the type array, just like before the `Unpack` call (`typesArray.Iterate(...`).

This also solves the problem of having to decide what to do with a decoded `*big.Int`

turbolent commented on fvm/evm/stdlib/contract.go:
```diff
@@ -38,6 +42,491 @@ var evmAddressBytesType = sema.NewConstantSizedType(nil, sema.UInt8Type, types.A
+						arrValues...,
+					)
+					values = append(values, arr)
+				}
```
Add a default case and report an error. I think that should never occur, so might have to be an implementation error, because the Cadence type to geth type conversion before the unpack should already  reject invalid types? 🤔 (not 100% sure)

turbolent commented on fvm/evm/stdlib/contract.go:
```diff
@@ -38,6 +42,491 @@ var evmAddressBytesType = sema.NewConstantSizedType(nil, sema.UInt8Type, types.A
+						panic(err)
+					}
+					arguments = append(arguments, gethABI.Argument{Type: typ})
+				}
```
Add a default case and report a user error that the given Cadence type is not supported for decoding. Currently, the type is silently ignored.

turbolent reviewed (commented):
👍

m-Peter commented:
@turbolent Much thanks for the top-notch reviews on this PR :raised_hands: They were really helpful.
I think I have addressed almost all of the comments, except for the arbitrary nesting of arrays, e.g. `[[UInt8]]`. For now I have added support only for arrays of leaf values. Later on, we can generalize this.
The code looks really procedural :sweat_smile: 
I am not sure how to improve it, given the nature of this desired functionality, and all the type-mapping needed. But I am open to suggestions :pray:

m-Peter commented:
https://cadence-lang.org/docs/language/values-and-types#fixed-point-numbers Given that Cadence has only `Fix64` & `UFix64`, maybe it is worth the effort to map them as well.

turbolent reviewed (commented):
Looks good! I'm refactoring this a bit more and will push the changes up soon

m-Peter commented:
@turbolent Very nice work on the refactoring :100: 
I have also added support for ABI encoding/decoding `[EVM.EVMAddress]`, fixed a bug in encoding/decoding of constant-size Cadence arrays and added relevant tests.

turbolent reviewed (approved):
Looks good! Thank you for the fixes and additional tests @m-Peter 👌 🙌 

Let's get this merged in, and then we can improve this in follow-up PRs (e.g. add computation metering)

ryan-suematsu-anchain commented:
@m-Peter thank you for all of your hard work!! took a look, it all looks good on my end.  As a final check could I get a list of all supported types so far?

turbolent commented:
@ryan-suematsu-anchain The currently supported types are:
- `string`
- `bool`
- `address`
- `uint8`, `uint16`, `uint32`, `uint64`, `uint128`, `uint256`
- `int8`, `int16`, `int32`, `int64`, `int128`, `int256`
- arrays of those types (recursive)

ryan-suematsu-anchain commented:
@turbolent that looks good!  I think this should cover everything we need.

ramtinms commented on fvm/environment/meter.go:
```diff
@@ -52,6 +52,9 @@ const (
 	ComputationKindRLPDecoding                = 2039
 	ComputationKindEncodeEvent                = 2040
+	_                                         = 2041
+	ComputationKindEVMEncodeABI               = 2042
+	ComputationKindEVMDecodeABI               = 2042
```
@m-Peter these two are reusing the same number. we might need different weights for them so we might use different numbers

ramtinms commented on fvm/evm/stdlib/contract.go:
```diff
@@ -38,6 +46,790 @@ var evmAddressBytesType = sema.NewConstantSizedType(nil, sema.UInt8Type, types.A
+			}
+
+			var arguments gethABI.Arguments
+			typesArray.Iterate(inter, func(element interpreter.Value) (resume bool) {
```
do we need any protection against depth attacks here?

turbolent replied:
See https://github.com/onflow/flow-go/pull/5146

ramtinms commented on fvm/environment/meter.go:
```diff
@@ -52,6 +52,9 @@ const (
 	ComputationKindRLPDecoding                = 2039
 	ComputationKindEncodeEvent                = 2040
+	_                                         = 2041
+	ComputationKindEVMEncodeABI               = 2042
+	ComputationKindEVMDecodeABI               = 2042
```
@m-Peter these two are reusing the same number. we might need different weights for them so we might use different numbers

turbolent replied:
See https://github.com/onflow/flow-go/pull/5146#discussion_r1427123276
